### ISSUE TITLE
Probability of applying a transform

### ISSUE THREAD

**Issue by agaldran** [automl/trivialaugment#6]
Hi Samuel,

Thanks for your fantastic work, even if TrivialAugment was a bit worse than all the other RL-based approaches, it should still be preferrable, given its very low cost and straightforward implementation, I think.

I am trying to understand your implementation in `auglib.py` in order to modify it to suit my purposes, and I am seeing that, in the `standard` search spaces, both `fixed` and `wide`, the geometric transforms inside `ALL_TRANSFORMS` (this is, rotate, translates, shears) all have a 50% chance of being applied. This is not true for the intensity-based transforms, though. I think (might be wrong) that this is not mentioned in the paper, e.g. in fig. 1 or the algorithm listing in page 4. So, may I ask if you used the 50% chance of switching off geometric transforms in your experiments?

Thanks!

Adrian

**Comment by SamuelGabriel** [comment]
Hi Adrian,

Thanks! You mean this if-statement for example: https://github.com/automl/trivialaugment/blob/3bfd06552336244b23b357b2c973859500328fbb/aug_lib.py#L122

In that case, I think you just misread the code. These if-statements are simply there such that the direction (left or right) in which a rotate/translate/shear happens is random. So the strength specifies the number of degrees and this randomness decides whether to go left or right. Does this answer your question?

**Comment by agaldran** [comment]
Hi,

Nah, it seems I got confused with the `if random.random() < probability:` inside the `TransformT` class definition, plus the comments that say `# extra coin-flip` inside the `set_augmentation_space` function. I can see now that you always use `probability=1`, and what you said about the sign of the transform. 

Now that I have you here, though, can I ask you if there is any difference between the `wide_standard` augmentations you provide here and the `trivial_augment_wide` augmentations in the official pytorch implementation that you contributed recently? The strengths in the augmentation space [here](https://github.com/pytorch/vision/blob/b5aa0915fe16e82ee4c24919032b4e7afae3ae1b/torchvision/transforms/autoaugment.py#L393) appear to be different, aren't they? In particular, for "Brightness", "Color", "Contrast", and "Sharpness", the values in `auglib.py` appear to vary roughly in [0,2] but in `torchvision` they are in [0,1], if I am reading correctly? 

Also, out of curiosity, may I ask why the `fixed_standard` option was not added into `torchvision`?

Cheers,

Adrian

**Comment by SamuelGabriel** [comment]
Hi Adrian,

There actually is a difference in augmentation space for some augmentations, the enhancer augmentations, but it is tiny, it is not in the range [.01,2], but in [.01,1.99]. This should not make any difference at all. :)
You think the ranges are very different, because you overlooked the signed option in torchvision 
https://github.com/pytorch/vision/blob/b5aa0915fe16e82ee4c24919032b4e7afae3ae1b/torchvision/transforms/autoaugment.py#L435
and the + 1 here
https://github.com/pytorch/vision/blob/b5aa0915fe16e82ee4c24919032b4e7afae3ae1b/torchvision/transforms/autoaugment.py#L58

So, while there is a difference, it is only tiny. Does this answer your concern? :)

Regarding adding `fixed_standard` to torchvision: I proposed to do so, but the team at torchvision decided to only add `wide_standard`. To be fair, it probably is the more interesting setting anyways, especially in the future where bigger models are more common.

**Comment by agaldran** [comment]
It does answer it, thank you so much for taking the time to answer :) 

By the way, I think I found some kind of funky behavior in torchvision geometric transforms that you may be interested in. First, as you of course know, in your library you deal only with PIL Image objects and their associated methods to transform images, like in this line:

https://github.com/automl/trivialaugment/blob/3bfd06552336244b23b357b2c973859500328fbb/aug_lib.py#L156

Now, in the torchvision implementation, the images can also be of tensor types, and for geometric stuff we will end up in `torchvision.transform.functional.py`, in the shear case here, to be exact:

https://github.com/pytorch/vision/blob/e3767f8aa51921460c89e1d396ccce097cebc9d5/torchvision/transforms/functional.py#L1078

See, if you shear an image using pil or skimage, apparently it behaves in a different way than if you shear them with `torchvision.transforms.functional`, as I am showing for a random image in the screenshot below (look at the left top corner of the images). 

![screenshot](https://user-images.githubusercontent.com/4498444/149410029-f8d3799f-5388-4f0b-9f9f-ba3b81f33668.png)



I don't know if this might lead to inconsistent results if you apply torchvision transforms on pil images or tensors, but I thought it could be interesting to know about it for you. The code for generating the screenshot is below. Thanks again for your attention! Cheers, Adrian

```
import numpy as np
import math
import matplotlib.pyplot as plt

from PIL import Image

from skimage import img_as_ubyte
from skimage.transform import AffineTransform, warp

from torchvision.transforms import functional as F, InterpolationMode

# load an image, call it im

# define shear magnitude. Take into account that:
# * torchvision.transform.functional expects shear angle value in clockwise direction in degrees [-180, 180]
# * skimage.transform.warp           expects shear angle in counter-clockwise direction as radians 
# * pil_img.transform(pil_img.size, Image.AFFINE, (a,b,c,d,e,f)) asks for radians, same as skimage
magnitude = 0.2

# shear_x as seen in torchvision/transforms/autoaugment.py 
im_tv = F.affine(im, angle=0.0, translate=[0, 0], scale=1.0, 
                 shear=[math.degrees(magnitude), 0.0],
                 interpolation=InterpolationMode.NEAREST, fill=0)

# shear_x as seen in automl/trivialaugment/aug_lib.py
im_pil = im.transform(im.size, Image.AFFINE, (1, magnitude, 0, 0, 1, 0))

# shear_x as in skimage
tf_mat = AffineTransform(np.array([[1, magnitude, 0], [0, 1, 0], [0, 0, 1]]))
im_sk = warp(np.array(im), inverse_map=tf_mat, order=0)
im_sk = img_as_ubyte(im_sk)

f, ax = plt.subplots(figsize=(16,6), nrows=1, ncols=3)
ax[0].imshow(im_tv)
ax[1].imshow(im_pil)
ax[2].imshow(im_sk)
_, _, _ = ax[0].axis('off'), ax[1].axis('off'), ax[2].axis('off')

ax[0].title.set_text('torchvision')
ax[1].title.set_text('pil')
ax[2].title.set_text('skimage')
```

**Comment by SamuelGabriel** [comment]
That really sounds interesting and like a problem for all of the implemented methods: RandAugment, AutoAugment and TrivialAugment :( Good find! I just tested it locally and checked again, it really seems to be the case, that the fix point is in different positions for PIL (top left/right) vs torchvision (center). Do you want to raise this in an issue on torchvision? If not, I would go ahead and do that. :)

I could reduce your example a code a little further:

```
from PIL import Image
import math
import torchvision # > 0.11

from torchvision.transforms import functional as F
interpolation = torchvision.transforms.InterpolationMode.NEAREST
fill = None

img = Image.new('RGB', (32,32), (255,255,0))
magnitude = .7

# shear_x as seen in torchvision/transforms/autoaugment.py 
im_torch = F.affine(img, angle=0.0, translate=[0, 0], scale=1.0, shear=[math.degrees(magnitude), 0.0],
                    interpolation=interpolation, fill=fill)

# shear_x as seen in automl/trivialaugment/aug_lib.py
im_pil = img.transform(img.size, Image.AFFINE, (1, magnitude, 0, 0, 1, 0))

im_torch.show()
im_pil.show()
```

**Comment by agaldran** [comment]
Hi,

Yes, it's quite surprising, and not a bug from the autoaugment functions, but an unexpected behavior of torchvision in my opinion. What the user would expect is that their `_get_inverse_affine_matrix` here:

https://github.com/pytorch/vision/blob/1feb6376186069595a428488b46ade6bee8829c2/torchvision/transforms/functional.py#L947

would behave in the same way as its analog does in pil, or skimage, but apparently it does not; visually it seems as if they were doing something weird with the center of the transform, but I did not dig further.

Please go ahead, I am not going to have the time to deal with this one. Thank you very much!

Adrian

**Comment by vfdev-5** [comment]
@agaldran @SamuelGabriel coming here from torchvision's issue and trying to figure out what would be an expected behaviour for shear op. Pillow does not provide any shear parametrization (user gives inverse affine matrix). Scikit-image does shearing with a single parameter (X-axis) with scaling as matrix is `[1, sin(theta); 0, cos(theta)]` and its determinant is not 1, so output area is not preserved. Matplotlib does something like `[1, tan(theta_x), tan(theta_y), 1]` and for single axis preserves the image scale. Scikit-image vs matplotplib: https://thomasaarholt.github.io/affine_transforms/affine.html#Shear

As for torchvision's implementation, it does rotation part keeping image center invariant such that doing 45 degrees rotation image is still in the frame and less fillcolored area outside (compare with rotation keeping top-left corner). The fact that shear is also centered was not intended, but I wonder why it could be worse (inconsistent, yes) in the same way trying to keep more image data inside the frame ? If you compare above images, fillcolor area is less important for torchvision vs PIL or scikit-learn. What do you think ?

I agree that adding a `center` option and a flag to be able to apply non-centered shear to torchvision affine would help to have certain consistency with others...

**Comment by SamuelGabriel** [comment]
Hey @vfdev-5 :) 

We raised this only due to the inconsistency. Interesting point, that shear is generally not as clearly defined as I imagined it to be. After looking at the sheared images, I agree, likely fixing the center is better than the top, as you say. I do not have time/capacity right now to test whether this makes any difference though (or to test whether it really is not worse), but I might come back to it in the coming weeks, as I thought about running a few more trainings with TrivialAugment. Yep, the `center` keyword for sure would allow consistency. But I am not sure, whether it is a good idea for the `autoaugment` code to change the `shear` op then, as @datumbox used that implementation successfully for trainings previously.

**Comment by datumbox** [comment]
> But I am not sure, whether it is a good idea for the autoaugment code to change the shear op then, as @datumbox used that implementation successfully for trainings previously.

@vfdev-5 added it in a BC manner so we are good. :)

**Comment by agaldran** [comment]
Hi,

Sorry for the late reply. From a standard torchvision's user perspective, I would like to have as few black pixels as possible :)

I also think the user should be aware that what the affine transforms do not behave as the analogous ones do in PIL, scikit-image, etc. The `center` thing sounds perfect to me. 

Thanks!!
